hnix-store-core/src/System/Nix/Internal/Hash.hs

```diff
@@ -53,6 +53,22 @@ data HashAlgorithm
     -- input 'HashAlgorithm' to the given number of bytes. See
     -- 'truncateDigest' for a description of the truncation algorithm.
 
+class HashProperties a
+ where
+  canonicalHashName :: a -> Text
+  canonicalHashLen :: a -> Int
+
+instance HashProperties HashAlgorithm
+ where
+  canonicalHashName SHA256 = "sha256"  -- SHA256 is the most used in Nix - so it matches first
+  canonicalHashName MD5    = "md5"
+  canonicalHashName SHA1   = "sha1"
+  canonicalHashName SHA512 = "sha512"
+  canonicalHashLen  SHA256 = 32
+  canonicalHashLen  MD5    = 16
+  canonicalHashLen  SHA1   = 20
+  canonicalHashLen  SHA512 = 64
```

Comment:
This makes the function partial because you cannot define it for Truncated. Requiring a Named algo does not work either.

We could split Truncated outside of HashAlgorithm type. That would also prevent the weird possibility to create a `Truncated (Truncated hash)`

Reply:
In C++ nix i swear the only point of truncated was the intermediate result when calculating store paths. I wonder if we can get rid of it entirely in both code bases

Reply:
Idea is that to move code closer&closer to Cryptonite interface. Library has this `canonicalHash{Name,Len}` info in instances.

Put truncated separately seems to be the minimal core compromise, it allows/would allow to use hash libraries directly and see where Truncated is really used.

Slipped the idea of SRI hash form to Cryptonite. Since that hash form is minimal.

The truncated hash form is a more rare broad sensitive question, do not think it should or would or be included in lib.

My plan is to take a response about SRI in Cryptonite, ask recommendations to extend the library, and then maybe for the SRI and Truncated hash create separate tiny packages. since they are legitimate but rarely used cases, there can be some other users for them.

Reply:
Oh, silly me. Truncated has a canonicalHashLen. It's just the truncation length :-)

Reply:
It would, but that does not matter to the software design, redefining the core `HashAlgorithm` type means redefining the whole library interface, so I thought and after our back&forth I ensured that Trunkated should be given separately.